### PR TITLE
ENH GH20601 raise error when pivot table's number of levels > int32

### LINKED ISSUE #20601: ERR: pivot_table when number of levels larger than int32 range
#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
dat = pd.DataFrame({'ind1':list(range(1337600))*2,'ind2':list(range(3040))*2*440,'count':[1]*2*1337600})
dat.pivot_table(index='ind1',columns='ind2',values='count',aggfunc='count')
```

#### Problem description

Above code raises the following error:

```
  File "..\pandas\core\reshape\reshape.py", line 144, in _make_selectors
    mask = np.zeros(np.prod(self.full_shape), dtype=bool)

ValueError: negative dimensions are not allowed
```

np.prod(self.full_shape) appears to be returning a negative value because the number of unique index combinations is larger than the largest int32 value.

If line 144 were changed to the following, the issue could be fixed:
```
mask = np.zeros(np.prod(self.full_shape,dtype=np.int64), dtype=bool)
```

#### Expected Output
1337600 x 3040 dataframe.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 37 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.22.0
pytest: 3.3.2
pip: 9.0.1
setuptools: 38.4.0
Cython: 0.27.3
numpy: 1.12.1
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.6.6
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.2
openpyxl: 2.4.10
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by anhqle**
Because the error of number of levels exceeding int32 can happen both in `pivot_table` and `unstack`, I catch the error as early as possible in both places.

Arguments for and against catching the error in `pivot_table`:
- For: `pivot_table` does aggregation before calling `unstack`, so it takes a while before the error is raised
- Against: `pivot_table` ultimately calls `unstack`, so it's not worth it to check for the error in both places

I'm happy to go with either way as you prefer.

- [X ] closes #20601
- [X ] tests added / passed
- [X ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
generally use the original PR and push to it. having a new one is confusing.

File: pandas/core/reshape/pivot.py [hunk 1]
```diff
@@ -29,6 +29,11 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
     index = _convert_by(index)
     columns = _convert_by(columns)
 
+    num_rows = data.reindex(index, axis='columns').shape[0]
```
Comment (jreback) [inline pandas/core/reshape/pivot.py]: you are doing extra work here (e.g. the reindex), can we not do the calculation directly?

File: pandas/core/reshape/pivot.py [hunk 2]
```diff
@@ -29,6 +29,11 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
...
+    num_columns = data.reindex(columns, axis='columns').shape[0]
```
Comment (jreback) [inline pandas/core/reshape/pivot.py]: also this is error is now in 2 places

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/pivot.py (hunk 1 above)
Comment (anhqle) [reply pandas/core/reshape/pivot.py]: Originally I was using `data.loc[:, index].shape[0]`, but pandas throws the "Indexing with a list with missing labels is Deprecated" warning. I use [`reindex` as it's the idiomatic approach recommended by the doc](http://pandas.pydata.org/pandas-docs/stable/whatsnew.html#indexing-with-a-list-with-missing-labels-is-deprecated).

I'm happy to make any change if you could point me to a better approach.

File: pandas/core/reshape/pivot.py [hunk 3]
```diff
@@ -79,6 +79,13 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
                 pass
         values = list(values)
 
+    num_rows = (data.reindex(columns=index).drop_duplicates().shape[0]
+                if index else 1)
+    num_cols = (data.reindex(columns=columns).drop_duplicates().shape[0]
+                if columns else 1)
```
Comment (anhqle) [inline pandas/core/reshape/pivot.py]: The goal here is to get an accurate size of the resulting pivot table, taking into account potential duplicates.

However I recognize that the performance hit may not be worth it given that this is such an edge case. It's up to you whether we should just raise the error within `unstack`.

**Comment by anhqle** [comment]
My apology about creating a new PR. I `pull --rebase upstream master` in the original PR and the diff got polluted by other commits. Am I supposed to `pull --rebase upstream master`, or the maintainers are in charge of the final merge into `master`?

Thanks again for your guidance -- I'm learning my way through the codebase and hope to be an active contributor in the future.

**Comment by jreback** [comment]
your u should rebase against upstream every time you push - it makes it so your code is not out of date wrt to master
but you don’t need to squash

**Comment by anhqle** [comment]
Do you think this PR is ready to be merged? Happy to do any additional work.

**Comment by jreback** [comment]
can you rebase

File: pandas/core/reshape/reshape.py [hunk 4]
```diff
@@ -126,6 +126,13 @@ def __init__(self, values, index, level=-1, value_columns=None,
         self.removed_level = self.new_index_levels.pop(self.level)
         self.removed_level_full = index.levels[self.level]
 
+        num_rows = np.max([index_level.size for index_level
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: can you add a comment on the checking here

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note, bug fixes in reshaping for 0.24.0

**Comment by jreback** [comment]
can you rebase

**Comment by jreback** [comment]
closing in favor of #23512

**Pull request closed without merging**